In [1]:
import pandas as pd
import numpy as np

In [27]:
matches=pd.read_csv('../matches.csv')
deliveries=pd.read_csv('../deliveries.csv')

In [3]:
matches.head()

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

#### 1) Who will win the match? RCB or KKR?
#### 2) What will the toss winning caption decide to do?

In [4]:
matches.columns

Index(['id', 'season', 'city', 'date', 'team1', 'team2', 'toss_winner',
       'toss_decision', 'result', 'dl_applied', 'winner', 'win_by_runs',
       'win_by_wickets', 'player_of_match', 'venue', 'umpire1', 'umpire2',
       'umpire3'],
      dtype='object')

In [5]:
matches['team1'].value_counts()

Mumbai Indians                 101
Kings XI Punjab                 91
Chennai Super Kings             89
Royal Challengers Bangalore     85
Kolkata Knight Riders           83
Delhi Daredevils                72
Rajasthan Royals                67
Sunrisers Hyderabad             63
Deccan Chargers                 43
Pune Warriors                   20
Gujarat Lions                   14
Rising Pune Supergiant           8
Kochi Tuskers Kerala             7
Rising Pune Supergiants          7
Delhi Capitals                   6
Name: team1, dtype: int64

In [6]:
royal='Royal Challengers Bangalore'
kings='Kings XI Punjab'
matches.loc[((matches['team1'] ==royal) | (matches['team1'] ==kings)) & ((matches['team2'] ==royal) | (matches['team2'] ==kings))]['winner'].value_counts()

Kings XI Punjab                12
Royal Challengers Bangalore    12
Name: winner, dtype: int64

Both won same no of matches against each other - 12

In [7]:
matches.loc[(matches['winner'] ==royal)]['winner'].value_counts()

Royal Challengers Bangalore    84
Name: winner, dtype: int64

In [8]:
matches.loc[(matches['winner'] ==kings)]['winner'].value_counts()

Kings XI Punjab    82
Name: winner, dtype: int64

kings won 82 and royal won 84 matches till now

In [9]:
len(matches.loc[((matches['team1'] ==royal) | (matches['team2'] ==royal))]['winner'])

180

In [11]:
matches.loc[(matches['winner'] ==royal)]['winner'].value_counts()[0]/len(matches.loc[((matches['team1'] ==royal) | (matches['team2'] ==royal))])

0.4666666666666667

royal won 46.66 % of all matches played

In [12]:
matches.loc[(matches['winner'] ==kings)]['winner'].value_counts()[0]/len(matches.loc[((matches['team1'] ==kings) | (matches['team2'] ==kings))])

0.4659090909090909

kings won 46.59% of all matches played

In [13]:
matches.loc[((matches['team1'] ==royal) | (matches['team1'] ==kings)) & ((matches['team2'] ==royal) | (matches['team2'] ==kings))]['toss_winner'].value_counts()

Kings XI Punjab                13
Royal Challengers Bangalore    11
Name: toss_winner, dtype: int64

In [14]:
matches.loc[((matches['team1'] ==royal) | (matches['team1'] ==kings)) & ((matches['team2'] ==royal) | (matches['team2'] ==kings))]['toss_decision'].value_counts()

field    17
bat       7
Name: toss_decision, dtype: int64

kings won 13 tosses and royal 11. 17 times filed option chosen while 7 times it is to bat

In [15]:
matches.loc[((matches['team1'] ==royal) | (matches['team1'] ==kings)) & ((matches['team2'] ==royal) | (matches['team2'] ==kings))].groupby('winner')['toss_decision'].value_counts()

winner                       toss_decision
Kings XI Punjab              field            8
                             bat              4
Royal Challengers Bangalore  field            9
                             bat              3
Name: toss_decision, dtype: int64

kings chose to filed 66.66 % times & royal chose to field 75% times against each other.

In [17]:
matches.loc[((matches['team1'] ==royal) | (matches['team1'] ==kings)) & ((matches['team2'] ==royal) | (matches['team2'] ==kings))].groupby('toss_winner')['winner'].value_counts()

toss_winner                  winner                     
Kings XI Punjab              Royal Challengers Bangalore    7
                             Kings XI Punjab                6
Royal Challengers Bangalore  Kings XI Punjab                6
                             Royal Challengers Bangalore    5
Name: winner, dtype: int64

the chance of winning for both the teams if they win the toss is 50-50

In [24]:
matches.loc[((matches['team1'] ==royal) | (matches['team1'] ==kings)) & ((matches['team2'] ==royal) | (matches['team2'] ==kings))].groupby('season')['winner'].value_counts()

season  winner                     
2008    Kings XI Punjab                2
2009    Kings XI Punjab                1
        Royal Challengers Bangalore    1
2010    Royal Challengers Bangalore    2
2011    Kings XI Punjab                1
        Royal Challengers Bangalore    1
2012    Kings XI Punjab                1
        Royal Challengers Bangalore    1
2013    Kings XI Punjab                2
2014    Kings XI Punjab                2
2015    Kings XI Punjab                1
        Royal Challengers Bangalore    1
2016    Royal Challengers Bangalore    2
2017    Kings XI Punjab                2
2018    Royal Challengers Bangalore    2
2019    Royal Challengers Bangalore    2
Name: winner, dtype: int64

From past 2 years, all the matches between them are won by royal. then it is likely that the coming match will be won by royal

#### 3) How many will Dinesh karthik score on the day of match?

In [25]:
deliveries.head()

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  ...  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0  ...         0   
1     2  DA Warner    S Dhawan  TS Mills              0  ...         0   
2     3  DA Warner    S Dhawan  TS Mills              0  ...         0   
3     4  DA Warner    S Dhawan  TS Mills              0  ...         0   
4     5  DA Warner    S Dhawan  TS Mills              0  ...         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs  player_dismissed dismissal_kind fielder  
0           0               NaN            NaN     NaN  
1           0               NaN            NaN     NaN  
2           4               NaN            NaN     NaN  
3           0               NaN            NaN     NaN  
4           2               NaN            NaN     NaN  

[5 rows x 21 columns]

In [33]:
deliveries.columns

Index(['match_id', 'inning', 'batting_team', 'bowling_team', 'over', 'ball',
       'batsman', 'non_striker', 'bowler', 'is_super_over', 'wide_runs',
       'bye_runs', 'legbye_runs', 'noball_runs', 'penalty_runs',
       'batsman_runs', 'extra_runs', 'total_runs', 'player_dismissed',
       'dismissal_kind', 'fielder'],
      dtype='object')

In [39]:
for val in deliveries['batsman']:
    if 'Karthik' in val:
        print(val)
        break;

KD Karthik


In [40]:
sum(deliveries[deliveries['batsman']=='KD Karthik']['total_runs'])

3854

Total runs score by Karthik is 3854 till now

In [42]:
sum(deliveries[deliveries['batsman']=='KD Karthik']['total_runs'])/deliveries[deliveries['batsman']=='KD Karthik']['match_id'].nunique()

23.790123456790123

Avg score of Karthik per match is 23.79

In [43]:
sum(deliveries[(deliveries['batsman']=='KD Karthik') & (deliveries['bowling_team']==royal)]['total_runs'])/deliveries[(deliveries['batsman']=='KD Karthik') & (deliveries['bowling_team']==royal)]['match_id'].nunique()

21.95

Avg score of Karthik per match against royal is 21.95 

In [44]:
sum(deliveries[(deliveries['batsman']=='KD Karthik') & (deliveries['bowling_team']==royal) & (deliveries['inning']==1)]['total_runs'])/deliveries[(deliveries['batsman']=='KD Karthik') & (deliveries['bowling_team']==royal) & (deliveries['inning']==1)]['match_id'].nunique()

22.375

Avg score of Karthik per match against royal in 1st innings is 22.375

In [45]:
sum(deliveries[(deliveries['batsman']=='KD Karthik') & (deliveries['bowling_team']==royal) & (deliveries['inning']==2)]['total_runs'])/deliveries[(deliveries['batsman']=='KD Karthik') & (deliveries['bowling_team']==royal) & (deliveries['inning']==2)]['match_id'].nunique()

21.666666666666668

Avg score of Karthik per match against royal in 2nd innings is 21.66

In [57]:
chk=set(deliveries[(deliveries['batsman']=='KD Karthik') & (deliveries['bowling_team']==royal)]['match_id'])
print(chk)

{129, 11153, 20, 280, 417, 167, 197, 459, 76, 209, 595, 7896, 104, 360, 620, 367, 495, 7922, 633, 383}


In [64]:
scores=[]
for val in chk:
    scores.append(sum(deliveries[(deliveries['batsman']=='KD Karthik') & (deliveries['bowling_team']==royal) & (deliveries['match_id']==val)]['total_runs']))
print(scores)

[13, 21, 1, 20, 44, 32, 18, 0, 2, 6, 52, 43, 6, 48, 2, 17, 1, 23, 29, 61]


In [65]:
max(scores),min(scores)

(61, 0)

max 61 runs against royal & least is 0

In [66]:
np.sort(scores)

array([ 0,  1,  1,  2,  2,  6,  6, 13, 17, 18, 20, 21, 23, 29, 32, 43, 44,
       48, 52, 61])

In [70]:
np.sort(scores).std()

18.67745967737583

In [71]:
np.median(np.sort(scores))

19.0

#### How many no balls will be bowled in match?

In [74]:
deliveries['noball_runs'].value_counts()

0    178364
1       698
2         9
5         6
3         1
Name: noball_runs, dtype: int64

In [75]:
(sum(deliveries[((deliveries['batting_team']==royal) | (deliveries['batting_team']==kings)) & ((deliveries['bowling_team']==royal) | (deliveries['bowling_team']==kings))]['noball_runs'])) #/deliveries[(deliveries['batsman']=='KD Karthik') & (deliveries['bowling_team']==royal) & (deliveries['inning']==2)]['match_id'].nunique()

29

Total no balls between them in all the matches is 29

In [79]:
sum(deliveries[((deliveries['batting_team']==royal) | (deliveries['batting_team']==kings)) & ((deliveries['bowling_team']==royal) | (deliveries['bowling_team']==kings))]['noball_runs'])/deliveries[((deliveries['batting_team']==royal) | (deliveries['batting_team']==kings)) & ((deliveries['bowling_team']==royal) | (deliveries['bowling_team']==kings))]['match_id'].nunique()

1.2083333333333333

On avg there is 1.2 no balls per match between them

In [80]:
for val in deliveries['batsman']:
    if 'Kohli' in val:
        print(val)
        break;

V Kohli


In [82]:
len(deliveries[(deliveries['batsman']=='V Kohli') & (deliveries['batsman_runs']==4)])

482

Total 482 4's in all matches

In [84]:
len(deliveries[(deliveries['batsman']=='V Kohli') & (deliveries['batsman_runs']==4) & (deliveries['bowling_team']==kings)])

74

Total 74 4's against kings team

In [85]:
len(deliveries[(deliveries['batsman']=='V Kohli') & (deliveries['batsman_runs']==4) & (deliveries['bowling_team']==kings)])/deliveries[(deliveries['batsman']=='V Kohli') & (deliveries['batsman_runs']==4) & (deliveries['bowling_team']==kings)]['match_id'].nunique()

3.5238095238095237

on Avg 3.52 4's against kings team per match